[Fix] RepairArticle 조회 쿼리를 이슈가 있는 게시물과 없는 게시물을 정확히 구분하여 페이징 처리

ghrltjdtprbs opened the pull request:
# ⭐️ [Fix] RepairArticle 조회 쿼리를 이슈가 있는 게시물과 없는 게시물을 정확히 구분하여 페이징 처리

## 🛠️ 변경 사항 :
 - enabled=true인 이슈가 있는 게시물과 없는 게시물을 정확히 구분하여 중복 조회 방지
 - 이슈가 없는 게시물에 대한 쿼리를 `repairArticle.issues.isEmpty().or(repairArticle.issues.any().enabled.isFalse())`로 수정
 - 이슈가 있는 게시물과 없는 게시물을 각각 조회하여 결합한 후 페이지네이션 적용
 - 전체 게시물 수를 계산하는 쿼리를 추가하여 정확한 페이지네이션 처리


## 💡 관련 이슈

- #186 

## 📃 개발 유형

- [ ] 버그 수정
- [ ] 새로운 기능 개발
- [x] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트


## ⏱️ 작업 기간

- 작업 시작일: (2024-06-12)
- 작업 종료일: (2024-06-12)

ghrltjdtprbs committed:
[fix] RepairArticle 조회 쿼리를 이슈가 있는 게시물과 없는 게시물을 정확히 구분하여 페이징 처리

변경 사항 :
 - enabled=true인 이슈가 있는 게시물과 없는 게시물을 정확히 구분하여 중복 조회 방지
 - 이슈가 없는 게시물에 대한 쿼리를 `repairArticle.issues.isEmpty().or(repairArticle.issues.any().enabled.isFalse())`로 수정
 - 이슈가 있는 게시물과 없는 게시물을 각각 조회하여 결합한 후 페이지네이션 적용
 - 전체 게시물 수를 계산하는 쿼리를 추가하여 정확한 페이지네이션 처리

asd42270 reviewed (approved):
고생하셨습니다! 재현님 옆에서 열심히 듣고 있었습니당

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/repository/RepairArticleQueryDslRepositoryImpl.java:
```diff
@@ -86,26 +87,35 @@ public Page<RepairArticle> findByOfficeIdAndProgressInAndCategoryOrderByIdDesc(
+                progressPredicate,
+                categoryPredicate);
+
+        long total = totalCountQuery.fetchCount();
```
카운트쪽은 기존방식이 더 나은것 같슴다!
지금도 문제는 없을것 같긴 합니다!
원래도 호불호가 갈릴 수 있는 부분이기도 합니다!
그리고 갯수만 필요한 경우에는 select를 좀 더 다듬어서 쓰면 성능적으로 좋을 수 있슴다
select(repairArticle.id).from(repairAtricle) 같은 방식으로도 사용 할 수 있습니다!

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/repository/RepairArticleQueryDslRepositoryImpl.java:
```diff
@@ -86,26 +87,35 @@ public Page<RepairArticle> findByOfficeIdAndProgressInAndCategoryOrderByIdDesc(
                 repairArticle.issues.any().enabled.isTrue())
-            .orderBy(repairArticle.id.desc());
+            .orderBy(repairArticle.id.desc())
+            .offset(pageable.getOffset())
```
오프셋을 여기 붙인건 논리적으로 좋은 선택인것 같슴다!

Galmaeki reviewed (approved):
고생하셨습니다!

ghrltjdtprbs committed:
[refactor] CustomProgress 조회 메서드 수정

변경 사항 :
 - CustomProgressRepository에 findAllByOrderByStartScheduleAsc 메서드 추가하여 startSchedule 기준으로 정렬된 결과 반환
 - GetRepairArticleService의 getCustomProgressList 메서드 수정
 - articleId를 사용하여 특정 RepairArticle의 CustomProgress 항목만 필터링
 - startSchedule 기준으로 정렬된 CustomProgress 목록 반환

ghrltjdtprbs committed:
[fix] RepairArticle 조회 및 정렬 수정

변경 사항 :
- `RepairArticleQueryDslRepositoryImpl`에서 `redDot` 상태를 고려하여 게시물을 조회하도록 수정:
  - `enabled=true`인 이슈가 있고 사용자가 체크하지 않은 경우 `redDot=true`로 설정
  - `redDot=true`인 게시물만 상단에 정렬
  - `redDot`이 아닌 게시물은 하단에 정렬

- `OwnerArticleListService`에서 `redDot` 상태를 설정하는 로직 추가:
  - 사용자 체크된 이슈 ID 목록을 기반으로 `redDot` 상태 결정
  - `redDot`이 true인 경우 해당 게시물을 상단에 정렬

- 기타 코드 정리 및 리팩토링:
  - `OwnerArticleListService`에 `getCheckedIssueIdsForCurrentUser` 메서드 추가
  - `findSavedArticleIds` 메서드를 통해 사용자가 저장한 게시물 ID 목록 조회
  - `OwnerRepairArticleResponseDto`로 매핑 시 `redDot` 상태 반영

asd42270 reviewed (approved):
다시 또 재현님과 얘기하면서 리뷰했슴다 고생하셨습니다~~

Galmaeki reviewed (approved):
고생많으셨슴다
